Integration test failure

### Version(s)

2.0.0

### Which components

server- Integration tests

### Tested With

_No response_

### What happened

I've got the same issue as #952 , the build log is attached.

My environment is as follows:
Apache Maven 3.8.7 (b89d5959fcde851dcb1c8946a785a163f14e1e29)
Maven home: C:\apache-maven-3.8.7
Java version: 17.0.6, vendor: Oracle Corporation, runtime: C:\Program Files\Java\jdk-17


Default locale: fr_FR, platform encoding: Cp1252
OS name: "windows 10", version: "10.0", arch: "amd64", family: "windows"

I recall being able to build Leshan a few months back on Windows and my environment is likely to be the same as then. The build succeeds on my WSL (most notable difference is Java 11 there).

Any thought?

### How to reproduce

Just run a mvn clean install

### Relevant Output

_No response_

[leshan-build-fail-SecurityTest.txt](https://github.com/eclipse-leshan/leshan/files/11903612/leshan-build-fail-SecurityTest.txt)

I fixed the kubernetes thing that was apparently written to my hosts file for no legitimate reason and reverted to the good old 127.0.0.1 localhost line. However there's still something going on with the way the InetSocketAddress looks up the address/hostname. With the additional logging you recommended:

Caused by: org.eclipse.californium.scandium.dtls.HandshakeException: Certificate chain could not be validated - server identity 127.0.0.1:127.0.0.1/127.0.0.1 does not match certificate [
[
  Version: V3
  Subject: CN=Server signed with Intermediate CA
  Signature Algorithm: SHA256withECDSA, OID = 1.2.840.10045.4.3.2

  Key:  Sun EC public key, 256 bits
  public x coord: 81741892632623131560020269223745815777694899637370162187782131386479660134752
  public y coord: 51913889454012516220131240810654165014666744529872007761037466049035318492215
  parameters: secp256r1 [NIST P-256,X9.62 prime256v1] (1.2.840.10045.3.1.7)
  Validity: [From: Thu Nov 19 15:47:55 CET 2020,
               To: Sat Oct 26 16:47:55 CEST 2120]
  Issuer: CN=Leshan intermediate CA
  SerialNumber: [    4a8c4577]

Certificate Extensions: 6
[1]: ObjectId: 2.5.29.35 Criticality=false
AuthorityKeyIdentifier [
KeyIdentifier [
0000: AD 31 B9 AF 93 72 64 92   5A D2 95 1E E5 40 AE 35  .1...rd.Z....@.5
0010: 56 9B 60 D7                                        V.`.
]
]

[2]: ObjectId: 2.5.29.19 Criticality=false
BasicConstraints:[
  CA:false
  PathLen: undefined
]

[3]: ObjectId: 2.5.29.37 Criticality=false
ExtendedKeyUsages [
  serverAuth
]

[4]: ObjectId: 2.5.29.15 Criticality=true
KeyUsage [
  DigitalSignature
  Key_Agreement
]

[5]: ObjectId: 2.5.29.17 Criticality=false
SubjectAlternativeName [
  DNSName: localhost
]

[6]: ObjectId: 2.5.29.14 Criticality=false
SubjectKeyIdentifier [
KeyIdentifier [
0000: E2 C9 90 A0 98 E5 43 FE   72 BC 2E 01 2B 7E 2D A5  ......C.r...+.-.
0010: 9C 70 9B 06                                        .p..
]
]

]
  Algorithm: [SHA256withECDSA]
  Signature:
0000: 30 45 02 20 23 D3 6E 77   BC F1 81 47 21 78 FB 04  0E. #.nw...G!x..
0010: DD CE 66 DA 3D C2 53 A3   37 B5 DC 96 39 48 EF 87  ..f.=.S.7...9H..
0020: F0 E9 89 AA 02 21 00 EC   70 8C 6E 5A B0 8D 9C C7  .....!..p.nZ....
0030: AB A4 52 08 31 88 56 A9   7E F9 BC 29 03 05 D6 2E  ..R.1.V....)....
0040: BD B1 CB 18 4B C1 C4                               ....K..

]

--- a/leshan-integration-tests/src/test/java/org/eclipse/leshan/integration/tests/util/LeshanTestClientBuilder.java
+++ b/leshan-integration-tests/src/test/java/org/eclipse/leshan/integration/tests/util/LeshanTestClientBuilder.java
@@ -17,6 +17,7 @@ package org.eclipse.leshan.integration.tests.util;
 
 import static org.eclipse.leshan.core.LwM2mId.OSCORE;
 
+import java.net.InetSocketAddress;
 import java.net.URI;
 import java.security.PrivateKey;
 import java.security.PublicKey;
@@ -436,11 +437,14 @@ public class LeshanTestClientBuilder extends LeshanClientBuilder {
     private URI getServerUri() {
         LwM2mServerEndpoint endpoint = server.getEndpoint(protocolToUse);
         URI serverUri = endpoint.getURI();
+        // we force usage of "localhost" as hostname to be sure that X.509 test works.
+        // see : https://github.com/eclipse-leshan/leshan/issues/1461#issuecomment-1631143202
         if (proxy != null) {
             // if server is behind a proxy we use its URI
-            return EndpointUriUtil.replaceAddress(serverUri, proxy.getClientSideProxyAddress());
+            return EndpointUriUtil.replaceAddress(serverUri,
+                    new InetSocketAddress("localhost", proxy.getClientSideProxyAddress().getPort()));
         } else {
-            return serverUri;
+            return EndpointUriUtil.replaceAddress(serverUri, new InetSocketAddress("localhost", serverUri.getPort()));
         }
     }
 }

